src/rust/engine/fs/store/src/lib.rs

```diff
@@ -267,11 +267,84 @@ impl RemoteStore {
           }
           local_store
             .store_bytes(entry_type, digest.hash, bytes, true)
-            .await?;
+                .await?;
+            }
+            Ok(())
+        })
+        .await
+    }
+
+    async fn download_digest_to_local_batch(
+        &self,
+        local_store: local::ByteStore,
+        digests: Vec<Digest>,
+    ) -> Result<(), StoreError> {
+        let mut large_files: Vec<Digest> = Vec::new();
+        let mut chunks: Vec<Vec<Digest>> = Vec::new();
+        let mut current_chunk: Vec<Digest> = Vec::new();
+        let mut current_size = 0;
+
+        for digest in digests {
+            let store_into_fsdb = ByteStore::should_use_fsdb(EntryType::File, digest.size_bytes);
+            if store_into_fsdb {
+                large_files.push(digest);
+            } else {
+                if current_size + digest.size_bytes >= 3000000 {
```

Comment:
iirc this needs to be below the 4mb default limit on the grpc call. Calculating the exact size will be similar to what happens in the batch upload code. Or maybe we can live with a less that perfect guess at the start (hardcoded constant)

Reply:
I starting doing this and realized that this will load the entire tree into memory (potentially many gigabytes), which isn't desirable. Similar to `load` I think I'll pass the local_store into `remote.rs` as a writer, but for batch operations. This way I can write to the store after every batch response.

The api signature on the provider will also reflect this, no longer directly returning bytes but something like `Result<HashMap<Digest, Result<bool, String>>, String>`